File: app/src/main/java/in/hridayan/ashell/fragments/setup/OnboardingItem3Fragment.java
package in.hridayan.ashell.fragments.setup;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import in.hridayan.ashell.R;
import in.hridayan.ashell.UI.DialogUtils;
import in.hridayan.ashell.UI.ToastUtils;
import in.hridayan.ashell.config.Const;
import in.hridayan.ashell.config.Preferences;
import in.hridayan.ashell.databinding.FragmentOnboardingItem3Binding;
import in.hridayan.ashell.shell.RootShell;
import in.hridayan.ashell.shell.ShizukuShell;
import in.hridayan.ashell.utils.HapticUtils;
import in.hridayan.ashell.utils.Utils;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import rikka.shizuku.Shizuku;

public class OnboardingItem3Fragment extends Fragment
    implements ShizukuShell.ShizukuPermCallback, RootShell.RootPermCallback {
  private static FragmentOnboardingItem3Binding binding;
  private ShizukuShell shizukuShell;
  private RootShell rootShell;

  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    binding = FragmentOnboardingItem3Binding.inflate(inflater, container, false);

    shizukuShell = new ShizukuShell(requireContext(), this);
    rootShell = new RootShell(requireContext(), this);

    binding.root.setOnClickListener(
        v -> {
          HapticUtils.weakVibrate(v);
          // we donot select the widget unless we get root permission
          binding.root.setSelected(false);
          // request root permission
          requestRootPermission();

          binding.shizuku.setSelected(false);
        });

    binding.shizuku.setOnClickListener(
        v -> {
          HapticUtils.weakVibrate(v);

          binding.shizuku.setSelected(false);
          // start permission check
          shizukuShell.startPermissionCheck();
          // request shizuku permission
          requestShizukuPermission();

          binding.root.setSelected(false);
        });

    return binding.getRoot();
  }

  // request root permission
  private void requestRootPermission() {
    ExecutorService executor = Executors.newSingleThreadExecutor();
    executor.execute(
        () -> {
          RootShell.refresh();
          if (!RootShell.isDeviceRooted()) {
            requireActivity().runOnUiThread(this::handleRootUnavailability);
          } else {
            rootShell.startPermissionCheck(); // Starts permission check
          }
        });
    executor.shutdown();
  }

  @Override
  public void onRootPermGranted() {
    requireActivity()
        .runOnUiThread(
            () -> {
              binding.root.setSelected(true);
              permGrantedToast();
              // set default local adb mode to root
              Preferences.setLocalAdbMode(Const.ROOT_MODE);
            });
  }

  // show this dialog if device is not rooted
  private void handleRootUnavailability() {
    binding.root.setSelected(false);
    DialogUtils.rootUnavailableDialog(requireContext());
  }

  // request shizuku permission
  private void requestShizukuPermission() {
    // Shizuku is not installed or running
    if (!Shizuku.pingBinder()) handleShizukuUnavailability();
    // Shizuku is running but havenot granted permission to aShellYou
    else if (!ShizukuShell.hasPermission()) Shizuku.requestPermission(0);
  }

  // this block executes immediately after permission is granted
  @Override
  public void onShizukuPermGranted() {
    requireActivity()
        .runOnUiThread(
            () -> {
              binding.shizuku.setSelected(true);
              permGrantedToast();
              // set default local adb mode to shizuku
              Preferences.setLocalAdbMode(Const.SHIZUKU_MODE);
            });
  }

  private void handleShizukuUnavailability() {
    binding.shizuku.setSelected(false);
    // Show dialog that shizuku is unavailable
    DialogUtils.shizukuUnavailableDialog(requireContext());
  }

  private void permGrantedToast() {
    ToastUtils.showToast(requireContext(), "Granted", ToastUtils.LENGTH_SHORT);
  }

  public static boolean isRootSelected() {
    return binding.root.isSelected();
  }

  public boolean isShizukuSelected() {
    return binding.shizuku.isSelected();
  }
}

File: app/src/main/java/in/hridayan/ashell/fragments/setup/OnboardingItem3Fragment.kt
package `in`.hridayan.ashell.fragments.setup

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import `in`.hridayan.ashell.config.Const
import `in`.hridayan.ashell.config.Preferences
import `in`.hridayan.ashell.databinding.FragmentOnboardingItem1Binding
import `in`.hridayan.ashell.databinding.FragmentOnboardingItem3Binding
import `in`.hridayan.ashell.shell.localadb.RootShell
import `in`.hridayan.ashell.shell.localadb.RootShell.RootPermCallback
import `in`.hridayan.ashell.shell.localadb.ShizukuShell
import `in`.hridayan.ashell.shell.localadb.ShizukuShell.ShizukuPermCallback
import `in`.hridayan.ashell.ui.ToastUtils
import `in`.hridayan.ashell.ui.dialogs.ErrorDialogs
import `in`.hridayan.ashell.utils.HapticUtils
import rikka.shizuku.Shizuku
import java.util.concurrent.Executors

class OnboardingItem3Fragment : Fragment(), ShizukuPermCallback,
    RootPermCallback {
    private lateinit var shizukuShell: ShizukuShell
    private lateinit var rootShell: RootShell
    private lateinit var binding: FragmentOnboardingItem3Binding

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentOnboardingItem3Binding.inflate(inflater, container, false)

        shizukuShell = ShizukuShell(requireContext(), this)
        rootShell = RootShell(requireContext(), this)

        binding.root.setOnClickListener { v: View? ->
            HapticUtils.weakVibrate(v)
            // we donot select the widget unless we get root permission
            binding.root.isSelected = false
            // request root permission
            requestRootPermission()
            binding.shizuku.isSelected = false
        }

        binding.shizuku.setOnClickListener { v: View? ->
            HapticUtils.weakVibrate(v)
            binding.shizuku.isSelected = false
            // start permission check
            shizukuShell.startPermissionCheck()
            // request shizuku permission
            requestShizukuPermission()
            binding.root.isSelected = false
        }

        return binding.getRoot()
    }

    // request root permission
    private fun requestRootPermission() {
        val executor = Executors.newSingleThreadExecutor()
        executor.execute {
            RootShell.refresh()
            if (!RootShell.isDeviceRooted()) {
                requireActivity().runOnUiThread { this.handleRootUnavailability() }
            } else {
                rootShell.startPermissionCheck() // Starts permission check
            }
        }
        executor.shutdown()
    }

    override fun onRootPermGranted() {
        requireActivity()
            .runOnUiThread {
                binding.root.isSelected = true
                permGrantedToast()
                // set default local adb mode to root
                Preferences.setLocalAdbMode(Const.ROOT_MODE)
            }
    }

    // show this dialog if device is not rooted
    private fun handleRootUnavailability() {
        binding.root.isSelected = false
        ErrorDialogs.rootUnavailableDialog(requireContext())
    }

    // request shizuku permission
    private fun requestShizukuPermission() {
        // Shizuku is not installed or running
        if (!Shizuku.pingBinder()) handleShizukuUnavailability()
        else if (!ShizukuShell.hasPermission()) Shizuku.requestPermission(0)
    }

    // this block executes immediately after permission is granted
    override fun onShizukuPermGranted() {
        requireActivity()
            .runOnUiThread {
                binding.shizuku.isSelected = true
                permGrantedToast()
                // set default local adb mode to shizuku
                Preferences.setLocalAdbMode(Const.SHIZUKU_MODE)
            }
    }

    private fun handleShizukuUnavailability() {
        binding.shizuku.isSelected = false
        // Show dialog that shizuku is unavailable
        ErrorDialogs.shizukuUnavailableDialog(requireContext())
    }

    private fun permGrantedToast() {
        ToastUtils.showToast(requireContext(), "Granted", ToastUtils.LENGTH_SHORT)
    }

    val isShizukuSelected: Boolean
        get() = binding.shizuku.isSelected

    companion object {
        @SuppressLint("StaticFieldLeak")
        private lateinit var binding: FragmentOnboardingItem3Binding
        val isRootSelected: Boolean
            get() = binding.root.isSelected
    }
}
